Automattic/wp-calypso
# Seller XP - add celebration modal when a "sell" user first publishes a product

#### Changes proposed in this Pull Request

* This PR adds behavior to the Full Site Editing plugin which conditionally displays a celebration modal for a user who started their site via the "Sell" flow, when they first publish content with a payments block in it.

Conditions for showing this modal, as I see them:
* User just finished saving
    * This is complete, making use of a series of selectors from Core Gutenberg which check for whether entities are in the process of being saved. 
    * This was somewhat complicated, as the Site Editor is basically an amalgamation of two editors (editing template and editing the page/post content within the template), so the status of both editors needs to be checked. For the page/post editor it is simpler. 
* Editor has not yet displayed modal once
    * Complete with a basic state value 
* User is a seller
    * Complete with a new `useSiteIntent` hook which fetches information about the user's intent from their onboarding experience.
* Content includes product block
    * Complete in a rather convoluted way that I may be able to refine further. Once again, the complication lay in the Site Editor, for which the built-in `core/block-editor` selectors do not work to look at the block content of the page, only the surrounding template. I was, however, able to get and check the raw content of the post and look for the block that way; in other contexts, I can make use of the `block-editor` selectors.
* User has not seen modal before for this site
    * Complete, using a custom API endpoint mimicking the one already existing at `apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wp-rest-wpcom-block-editor-first-post-published-modal-controller.php`.

![image](https://user-images.githubusercontent.com/13437011/156636892-567d4ad8-ed90-4375-918e-1d6f951bb097.png)

#### Testing instructions



* Check out this PR to your local Calypso install.
* In your command line, navigate to `apps/editing-toolkit/` and run `yarn dev --sync` to build and sync the changes from this PR to your sandbox.
    * See PCYsg-ly5-p2 for info on how to set your SSH connection up to your sandbox to make this work, if it is not already prepped for ETK work.
* Open the Site Editor for a site you have sandboxed, and verify that editing and then saving the editor (both editing the surrounding template and the page content itself, which are two different events according to `core/data`) triggers the modal to appear if the page content has a payments block in it and does nothing on subsequent reloads.
* Verify that if the page content does *not* have a payments block, that this does not happen even on the first save for a site.
* Navigate to the page editor for a page on the site, and the post editor for a post on the site, and repeat the above test, verifying that the modal only appears if the content includes a payment block.

**Note that the intention here is that the modal will only ever appear once per site; we set a blog option at the same time that we display it, `has_seen_seller_celebration_modal`. If you have seen the modal once on a particular site, you can manually reset this blog option using `wp shell`:
```
update_blog_option({id_of_your_sandboxed_test_site}, "has_seen_seller_celebration_modal", false)
```
to simulate a blog meeting the conditions for displaying the modal for the first time.**



Fixes #60413
